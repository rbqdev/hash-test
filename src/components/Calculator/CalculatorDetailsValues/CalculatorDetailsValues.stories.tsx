import { Story, Meta } from "@storybook/react/types-6-0";
import CalculatorDetailsValues, { ICalculatorDetailsValues } from ".";

export default {
  title: "CalculatorDetailsValues",
  component: CalculatorDetailsValues,
  args: {
    values: {
      1: 1,
      15: 15,
      30: 30,
      90: 90,
      120: 120
    }
  }
} as Meta;

export const Default: Story<ICalculatorDetailsValues> = args => (
  <CalculatorDetailsValues {...args} />
);
