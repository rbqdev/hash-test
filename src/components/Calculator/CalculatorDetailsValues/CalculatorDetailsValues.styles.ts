import styled from "styled-components";

export const List = styled.ul`
  display: flex;
  flex-direction: column;
  padding: 0;
`;

export const ListItem = styled.li`
  color: ${({ theme }) => theme.colors.blueLight};
  display: flex;
  line-height: 46px;

  > * {
    white-space: nowrap;
  }

  h3 {
    font-weight: normal;
    font-size: 1rem;
    margin: 0;
  }

  strong {
    margin-left: 3px;
  }
`;
