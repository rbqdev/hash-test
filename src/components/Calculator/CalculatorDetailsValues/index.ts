export { default } from "./CalculatorDetailsValues";
export type { ICalculatorCardValues } from "./CalculatorDetailsValues";
