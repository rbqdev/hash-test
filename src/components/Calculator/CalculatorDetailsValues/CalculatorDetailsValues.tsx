import Skeleton, { SkeletonTheme } from "react-loading-skeleton";
import formatNumber, { currencyConfigs } from "@utils/formatNumber";
import getListOfDays from "@utils/getListOfDays";
import getTextDay from "@utils/getTextDay";
import { _commonColors } from "@styles/themes";
import * as Styled from "./CalculatorDetailsValues.styles";

export interface ICalculatorDetailsValues {
  values: { [key: number]: number | string };
  isLoading: boolean;
}

export default function CalculatorDetailsValues({
  values = {},
  isLoading = false
}: ICalculatorDetailsValues) {
  const keysValues = Object.keys(values);
  const hasValues = keysValues.length > 0;
  const newListOfDays = keysValues.map(value => Number(value));
  const listOfDays = hasValues ? newListOfDays : getListOfDays();

  return (
    <Styled.List>
      {listOfDays.map(day => (
        <Styled.ListItem key={day}>
          {isLoading ? (
            <SkeletonTheme
              color={_commonColors.info}
              highlightColor={_commonColors.grey}
            >
              <Skeleton width={140} />
            </SkeletonTheme>
          ) : (
            <>
              <h3>{getTextDay(day)}:</h3>
              <strong>
                {formatNumber({
                  value: `${values?.[day] ?? "0"}`,
                  ...currencyConfigs
                })}
              </strong>
            </>
          )}
        </Styled.ListItem>
      ))}
    </Styled.List>
  );
}
