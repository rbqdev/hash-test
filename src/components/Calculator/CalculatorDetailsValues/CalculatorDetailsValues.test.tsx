import render from "@testsUtils/renderWithTheme";
import CalculatorDetailsValues from ".";

const renderComponent = ({ values = {}, isLoading = false }: any = {}) =>
  render(<CalculatorDetailsValues values={values} isLoading={isLoading} />);

describe("Component: CalculatorDetailsValues", () => {
  it("should render component correctly", () => {
    const { baseElement } = renderComponent();
    expect(baseElement).toMatchSnapshot();
  });

  describe("if prop 'isLoading'", () => {
    it("is `true`, should render squeleton elements", () => {
      const { container } = renderComponent({ isLoading: true });
      const elementExpected = container.querySelector(
        ".react-loading-skeleton"
      );

      expect(elementExpected).toBeInTheDocument();
    });

    it("is `false`, should NOT render squeleton elements", () => {
      const { container } = renderComponent();

      expect(
        container.querySelector(".react-loading-skeleton")
      ).not.toBeInTheDocument();
    });

    const mockTextAmanha = "AmanhÃ£:";
    const mockText_15_dias = "Em 15 dias:";
    const mockCurrency_0 = "R$ 0,00";

    test.each([
      [mockTextAmanha, mockCurrency_0, {}],
      [mockText_15_dias, mockCurrency_0, {}],
      [mockTextAmanha, mockCurrency_0, { 1: 0 }],
      [mockText_15_dias, mockCurrency_0, { 15: 0 }],
      [mockTextAmanha, "R$ 20,92", { 1: 2092 }],
      [mockText_15_dias, "R$ 21,12", { 15: 2112 }]
    ])(
      "should render text: '%s' with currency: '%s', getting data: '%s'",
      (textExpected, currencyExpected, values) => {
        const { getByRole, getAllByText } = renderComponent({ values });

        expect(
          getByRole("heading", {
            name: textExpected
          })
        ).toBeInTheDocument();

        expect(getAllByText(currencyExpected)[0]).toBeInTheDocument();
      }
    );
  });
});
