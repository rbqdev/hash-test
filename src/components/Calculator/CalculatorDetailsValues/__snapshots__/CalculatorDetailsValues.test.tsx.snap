// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: CalculatorDetailsValues should render component correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 0;
}

.c1 {
  color: #5d9cec;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  line-height: 46px;
}

.c1 > * {
  white-space: nowrap;
}

.c1 h3 {
  font-weight: normal;
  font-size: 1rem;
  margin: 0;
}

.c1 strong {
  margin-left: 3px;
}

<body>
  <div>
    <ul
      class="c0"
    >
      <li
        class="c1"
      >
        <h3>
          Amanhã
          :
        </h3>
        <strong>
          R$ 0,00
        </strong>
      </li>
      <li
        class="c1"
      >
        <h3>
          Em 15 dias
          :
        </h3>
        <strong>
          R$ 0,00
        </strong>
      </li>
      <li
        class="c1"
      >
        <h3>
          Em 30 dias
          :
        </h3>
        <strong>
          R$ 0,00
        </strong>
      </li>
      <li
        class="c1"
      >
        <h3>
          Em 90 dias
          :
        </h3>
        <strong>
          R$ 0,00
        </strong>
      </li>
    </ul>
  </div>
</body>
`;
